### PR TITLE
ci: split tests for local / db [vol.2]

### PR DESCRIPTION
**Pull request by Borda**
copy of #142, see: https://github.com/python-cachier/cachier/pull/142#issuecomment-1921522424

cc: @shaypal5

### CODE REVIEW

**Comment by shaypal5** [comment]
Thank you! Sorry for the bother.

**Comment by Borda** [comment]
> Thank you! Sorry for the bother.

no problem, it can happen :rabbit:

**Comment by shaypal5** [comment]
pytest seems to be angry with the "NOT mongo".
The mongo marker is declared as required.
Maybe a lowercase "not"?
@Borda

**Comment by Borda** [comment]
> pytest seems to be angry with the "NOT mongo".
> The mongo marker is declared as required.
> Maybe a lowercase "not"?

not sure I was using NOT without any issue, lets try it :)

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -31,15 +32,18 @@ jobs:
         run: |
           python -m pip install --upgrade pip
           python -m pip install -e . -r tests/requirements.txt
-      - name: Unit tests
+      - name: Unit tests (local)
+        if: matrix.backend == 'local'
+        run: pytest -m "NOT mongo"
```
Comment (Borda) [inline .github/workflows/test.yml:37]:
```suggestion
        run: pytest -m "not mongo"
```

**Comment by shaypal5** [comment]
This is the result of several bugs, one my own and the others... somebody's (perhaps mine; I didn't bother to check the git blame):
1. My own: The `_test_mongetter()` fixture breaks if a certain environment variables.
2. One test in `test_defaults.py`, named `test_mongetter_default_param()`, wasn't marked with the `mongo` mark.
3. A couple of generalized tests in `test_general.py` are parameterized in a way that tests all three cores. 

I fixed all three:
1. The missing environment variable is now handled gracefully.
2. `test_mongetter_default_param()` is now marked with `pytest.mark.mongo`.
3. No nice solution here (without digging into more complicated options). I didn't want to mark it with the `mongo` mark and thus "hide" a couple of important general tests in the mongo suite. I opted for duplicating the test (another bad solution, yes), and its mongo version is now in the `test_mongo_core.py`, and is marked properly. Perhaps a solution avoiding duplicate code can be found at a later date.

All three solutions can be found in PR #144 .
Would you mind reviewing it?

Cheers,
Shay

**Comment by Borda** [comment]
> All three solutions can be found in PR #144 .

yes, that looks good

**Comment by shaypal5** [comment]
Ok, everything passed, so I just merged it.
You can now rebase over master to continue working - hopefully this will solve the issues you've stumbled upon.

**Comment by Borda** [comment]
@shaypal5 seems all local pas and only DB is failing

**Comment by Borda** [comment]
@shaypal5 seems some locals are still failing, mind have look, also feel free to push changes directly to this PR :chipmunk:

**Comment by shaypal5** [comment]
Ok, we're upping our game:
1. I've pulled this into an origin dev branch with the name `split-tests`.
2. I've invited you to the cachier organization and to the group of collaborators, which will give you a write access to this repo.
3. I've protected the master branch, requiring an approved PR to merge into it (and disabling pushes to it, etc.).

I wanted to require some checks to pass before merging, and when trying to select one from the list, it made me realize the current matricization of the the test job across different platforms make Github consider them as a separate checks, which is what has been annoying you in the PR conversation. The upside is that all such builds run in parallel. Please take this into consideration when suggesting an alternative; perhaps this can be collapsed somehow just for the conversation view's sake.

Anyway, from now on, please work directly on dev branches of this repo and not your fork; you'll now be able to get your contributions tests directly against the live MongoDB server I've set up on the MongoDB Atlas platform on every push to your dev branches! :)

**Comment by shaypal5** [comment]
Let's port it.
And there's already a PR waiting for your approval. :)

**Pull request merged**